lib/fsm/utils.go

```diff
@@ -41,9 +41,54 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
 		return trace.BadParameter(
 			"phase %q has already been rolled back", phase.ID)
 	}
+
+	leafPhases := plan.GetLeafPhases()
+
+	// TODO: Rollback of non-leaf phases may currently not be implemented
+	// properly. Roll back starts top-down, and not in reverse order.
+	if !containsPhase(leafPhases, phaseID) {
+		return trace.BadParameter(
+			"attempting to rollback non-leaf phase %q", phase.ID)
+	}
+	if !latestRollback(leafPhases, phaseID) {
+		return trace.BadParameter(
+			"rollback subsequent phases before rolling back phase %q", phase.ID)
+	}
 	return nil
 }
 
+// containsPhase returns true if phases contains a phase with ID matching
+// phaseID.
+func containsPhase(phases []storage.OperationPhase, phaseID string) bool {
+	for _, phase := range phases {
+		if phase.ID == phaseID {
+			return true
+		}
+	}
+	return false
+}
+
+// latestRollback returns true if the phase specified by phaseID is next in line
+// to be rolled back.
+func latestRollback(phases []storage.OperationPhase, phaseID string) bool {
```

Comment:
I'm a bit worried about this algorithm to determine dependence, especially since we kind of support parallel phase execution which I'm probably going to turn on for upgrades of large clusters. By looking at the latest phase from the flattened structure, it creates a dependency that doesn't necessarily exist between phases. IE if in an upgrade I want to rollback / retry bootstrap on node-1, I have to rollback nodes 2-800. 

I think identifying when a rollback can happen would be a closer fit if any depends on operations are in progress, failed, completed. So it's the same readiness to move forwards as it is to move backwords. Although this is a much much more complicated algorithm with the way dependencies are structures, because they can point to a group phase or leaf phase.

Reply:
> IE if in an upgrade I want to rollback / retry bootstrap on node-1, I have to rollback nodes 2-800.

This is exactly the situation that we want to prevent customers from getting in, unless they know what they're doing. In a recent support scenario, someone rolled back one of masters without rolling back any of subsequent phases which put cluster in a bad state. For "power users" there should be a --force flag to override this check and allow you to rollback anything. @bernardjkim - btw I don't see this flag implemented?

Reply:
Nevermind, I see the --force flag is handled above.

Reply:
@r0mant @knisbet 
I've updated the implementation in 0bf4934 to be more accommodating towards rolling back parallel phases.